File: DSA/Ayush_Mishra_590014844/Day_58_Question_3.c
#include <stdio.h>
void maxHeapify(int arr[], int i, int n) {
    int largest = i;
    int left = 2*i + 1;
    int right = 2*i + 2;

    if (left < n && arr[left] > arr[largest])
        largest = left;
    if (right < n && arr[right] > arr[largest])
        largest = right;

    if (largest != i) {
        int temp = arr[i];
        arr[i] = arr[largest];
        arr[largest] = temp;
        maxHeapify(arr, largest, n);
    }
}

void replaceRoot(int arr[], int n, int newElement) {
    arr[0] = newElement;
    maxHeapify(arr, 0, n);
}

int main() {
    int n, newElement;
    scanf("%d", &n);
    int arr[n];
    for (int i = 0; i < n; i++)
        scanf("%d", &arr[i]);
    scanf("%d", &newElement);

    replaceRoot(arr, n, newElement);
    for (int i = 0; i < n; i++)
        printf("%d ", arr[i]);
    printf("\n");
    return 0;
}

File: DSA/VenikaPanwar_590013755/Day_58_Question_3.java
import java.util.*;

public class ReplaceRootInHeap {
    public static void replaceRoot(int[] heap, int newElement) {
        if (heap.length == 0) return;

        heap[0] = newElement; // replace root
        heapify(heap, heap.length, 0);
    }

    private static void heapify(int[] arr, int n, int i) {
        int largest = i;
        int left = 2 * i + 1;
        int right = 2 * i + 2;

        if (left < n && arr[left] > arr[largest]) largest = left;
        if (right < n && arr[right] > arr[largest]) largest = right;

        if (largest != i) {
            int temp = arr[i];
            arr[i] = arr[largest];
            arr[largest] = temp;

            heapify(arr, n, largest);
        }
    }

    public static void main(String[] args) {
        int[] heap1 = {50, 30, 20, 15, 10, 8, 16};
        replaceRoot(heap1, 25);
        System.out.println(Arrays.toString(heap1));

        int[] heap2 = {100, 80, 70, 60, 50, 40, 30};
        replaceRoot(heap2, 35);
        System.out.println(Arrays.toString(heap2));
    }
}
